src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1482,6 +1492,9 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
             newIdToTrackerMapBuilder[projectId] = newTracker;
         }
 
+        // Update _projectIdToTrackerMap to include all the newly created compilation trackers.
+        RoslynImmutableInterlocked.InterlockedExchange(ref _projectIdToTrackerMap, updatedIdToTrackerMapBuilder.ToImmutable());
```

Comment:
i feel like there's some sort of race here.   where multiple threads come in, update the build, and the other maps, and then blast back into this field.  this actually feels like it could potentially be a problem.  convince me this is 100% safe.

Reply:
this may need to be a spin, with an interlocked compare-exchange.  and if the field changes out from underneath us, we have to go redo this work to pick up any comp trackers made by another thread.

Reply:
Actually, I'm now not convinced there wasn't a timing issue in this code before.

Reply:
Actually, this PR was the one that added the multiple usages of _projectIdToTrackerMap within the method. I've pushed out a commit that I *think* essentially does what you requested.